### PR TITLE
RavenDB-22457 Next Gen Pagers - Part II

### PR DESCRIPTION
**Pull request by ayende**
More work around 6.2 & performance

https://issues.hibernatingrhinos.com/issue/RavenDB-22457

### CODE REVIEW

**Comment by ppekrol** [comment]
@ayende can you check failing tests?

**Comment by redknightlois** [comment]
Can we get this one merged? Errors doesn't seem to be related to changes happening in this set and my PRs (all 3) requires this to pass green :D

File: src/Voron/Util/ActiveTransactions.cs [hunk 1]
```diff
@@ -57,9 +57,26 @@ public bool TryRemove(LowLevelTransaction tx)
...
+            long latestReadTransactionId = tx.Environment.CurrentStateRecord.TransactionId;
+            if (latestReadTransactionId == oldTx)
                 return true;
 
+            if (tx.Flags is TransactionFlags.ReadWrite)
+            {
+                if (_activeTxs.HasTransactions == false)
+                {
+                    // To be safe, we publish the latest _read_ transaction, because we aren't done committing the current one yet 
+                    Interlocked.CompareExchange(ref _oldestTransaction, latestReadTransactionId, oldTx);
```
Comment (redknightlois) [inline src/Voron/Util/ActiveTransactions.cs:73]: Can you add some more detailed information regarding why that happens and in what scenario?

Comment (ayende) [reply src/Voron/Util/ActiveTransactions.cs:73]: Yes, updated the comment

**Comment by arekpalinski** [comment]
test this in debug please

File: src/Voron/Util/ActiveTransactions.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Voron/Util/ActiveTransactions.cs:73]: As I understand we'll have the oldest transaction bumped earlier (while previously it was always "actual oldest - 1" - which was safe approach).

What is the motivation for doing this change? What do we want to achieve?

Comment (redknightlois) [reply src/Voron/Util/ActiveTransactions.cs:73]: @arekpalinski It is clearly an optimization, and I understand what it is been done. However, I have to agree with you that I don't think that I can grasp the exact scenario either.

**Comment by redknightlois** [comment]
@arekpalinski the errors found are not related to this PR. Would be great if you can push this to v6.2 in order for me to have it available to rebase https://github.com/ravendb/ravendb/pull/19039 on top of this.

File: src/Voron/Util/ActiveTransactions.cs (hunk 1 above)
Comment (ayende) [reply src/Voron/Util/ActiveTransactions.cs:73]: `TryRemove` would scan all the active transactions whenever a transaction was disposed.
That isn't needed when we know that we are the latest anyway, so we changed it.

However, for _write_ transactions, we need to increment the oldest transaction. This is done here by moving to the latest published one, if there are no read tx.

There is no actual change in behavior, though.

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Client/Util/StreamWithTimeout.cs [hunk 2]
```diff
@@ -2,110 +2,32 @@
...
         private CancellationTokenSource _writeCts;
         private CancellationTokenSource _readCts;
 
         private long _totalRead = 0;
         private long _totalWritten = 0;
 
-        public StreamWithTimeout(Stream stream)
-        {
-            _stream = stream;
-            SetWriteTimeoutIfNeeded(DefaultWriteTimeout);
```
Comment (grisha-kotler) [inline src/Raven.Client/Util/StreamWithTimeout.cs:28]: You removed the timeout for all requests.
On `RavenDB.Client` we set the timeout to 12 hours.

- We should still set the timeout for old clients.
- Remove it from the new one.
- Make sure that there is timeout in Kestrel?

Comment (ppekrol) [reply src/Raven.Client/Util/StreamWithTimeout.cs:28]: I'm talking with @ayende about that tomorrow

**Pull request merged**